cdc/sorter/leveldb/table_sorter.go

```diff
@@ -0,0 +1,514 @@
+	"github.com/pingcap/ticdc/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	lutil "github.com/syndtr/goleveldb/leveldb/util"
+	"go.uber.org/zap"
+)
+
+const (
+	// Capacity of leveldb sorter input and output channels.
+	sorterInputCap, sorterOutputCap = 128, 128
+	// Max size of received event batch.
+	batchReceiveEventSize = 64
+)
+
+var levelDBSorterIDAlloc uint32 = 0
+
+func allocID() uint32 {
+	return atomic.AddUint32(&levelDBSorterIDAlloc, 1)
+}
+
+// Sorter accepts out-of-order raw kv entries and output sorted entries
+type Sorter struct {
+	actorID actor.ID
+	router  *actor.Router
+	uid     uint32
+	tableID uint64
+	serde   *encoding.MsgPackGenSerde
+
+	lastSentResolvedTs uint64
+	lastEvent          *model.PolymorphicEvent
+
+	inputCh  chan *model.PolymorphicEvent
+	outputCh chan *model.PolymorphicEvent
+
+	closed int32
+
+	metricTotalEventsKV         prometheus.Counter
+	metricTotalEventsResolvedTs prometheus.Counter
+}
+
+// NewLevelDBSorter creates a new LevelDBSorter
+func NewLevelDBSorter(
+	ctx context.Context, tableID int64, startTs uint64,
+	router *actor.Router, actorID actor.ID,
+) *Sorter {
+	captureAddr := util.CaptureAddrFromCtx(ctx)
+	changefeedID := util.ChangefeedIDFromCtx(ctx)
+	return &Sorter{
+		actorID:            actorID,
+		router:             router,
+		uid:                allocID(),
+		tableID:            uint64(tableID),
+		lastSentResolvedTs: startTs,
+		serde:              &encoding.MsgPackGenSerde{},
+
+		inputCh:  make(chan *model.PolymorphicEvent, sorterInputCap),
+		outputCh: make(chan *model.PolymorphicEvent, sorterOutputCap),
+
+		metricTotalEventsKV:         sorter.EventCount.WithLabelValues(captureAddr, changefeedID, "kv"),
+		metricTotalEventsResolvedTs: sorter.EventCount.WithLabelValues(captureAddr, changefeedID, "resolved"),
+	}
+}
+
+// waitInputOutput wait input or output becomes available.
+// It returns
+//   1) a max commit ts of received new events,
+//   2) a max resolved ts of new resolvedTs events,
+//   3) number of received new events,
+//   4) error.
+//
+// If input is available, it batch receives new events.
+// If output available, it sends a resolved ts event and returns.
+func (ls *Sorter) waitInputOutput(
+	ctx context.Context,
+	events []*model.PolymorphicEvent, outputBuf *outputBuffer,
+) (uint64, uint64, int, error) {
+	batchSize := len(events)
+	if batchSize <= 0 {
+		log.Panic("batch size must be larger than 0")
+	}
+	maxCommitTs, maxResolvedTs := uint64(0), uint64(0)
+	inputN := 0
+	kvEventCount, resolvedEventCount := 0, 0
+	appendInputEvent := func(ev *model.PolymorphicEvent) {
+		if ev.RawKV.OpType == model.OpTypeResolved {
+			if maxResolvedTs < ev.CRTs {
+				maxResolvedTs = ev.CRTs
+			}
+			resolvedEventCount++
+		} else {
+			if maxCommitTs < ev.CRTs {
+				maxCommitTs = ev.CRTs
+			}
+			events[inputN] = ev
+			inputN++
+			kvEventCount++
+		}
+	}
+
+	outputN := 0
+	outputEvent := model.NewResolvedPolymorphicEvent(0, ls.lastSentResolvedTs)
```

Comment:
so the regionID is not used? kinds of debug info?

Reply:
Yes, it's debug info and okay to be 0.